third_party/amd/lib/TritonAMDGPUTransforms/MfmaGroup.cpp

```diff
@@ -1,387 +1,285 @@
+  } else if (version <= 3 && isa<Float8E5M2Type>(aET) &&
+             isa<Float8E5M2Type>(bET)) {
+    // For the OCP FP8 E5M2 type, we can emulate the support for it with FP16.
+    aElemType = bElemType = aET = bET = b.getF16Type();
   }
-  if (llvm::isa<Float8E5M2FNUZType>(dataTypeA) &&
-      llvm::isa<Float8E5M2FNUZType>(dataTypeB)) {
-    return MfmaTypeId::Bf8Bf8TyId;
+  return {version, mDim, nDim, aET.getTypeID(), bET.getTypeID()};
+}
+
+//===----------------------------------------------------------------------===//
+// MFMA intrinsic map
+//===----------------------------------------------------------------------===//
+
+using MfmaMapValue =
+    std::tuple<StringRef /*symbol*/, unsigned /*kDim*/, unsigned /*kBase*/>;
+using MfmaMap = llvm::DenseMap<MfmaKey, SmallVector<MfmaMapValue, 2>>;
+
+class MfmaDatabase {
+public:
+  static const MfmaMap &get(MLIRContext *context) {
+    static MfmaDatabase db(context);
+    return db.mfmaMap;
   }
-  if (llvm::isa<Float8E5M2Type>(dataTypeA) &&
-      llvm::isa<Float8E5M2Type>(dataTypeB)) {
-    return MfmaTypeId::Fp16TyId;
+
+private:
+  explicit MfmaDatabase(MLIRContext *context);
+
+  MfmaMap mfmaMap;
+};
+
+MfmaDatabase::MfmaDatabase(MLIRContext *context) {
+// Macro for defining MFMA intrinsics at a specific gfx version.
+#define TRITON_MFMA_v(v, m, n, aET, bET, symbol, k, kBase)                     \
+  {                                                                            \
+    /*key=*/{v, m, n, aET.getTypeID(), bET.getTypeID()}, /*value=*/{           \
+      {ROCDL::symbol::getOperationName(), k, kBase},                           \
+    }                                                                          \
   }
-  if (isF8F6F4(dataTypeA) && isF8F6F4(dataTypeB)) {
-    return MfmaTypeId::F8F6F4TyId;
+// For gfx950, we can have two intrinsics with the same M/N but different K.
+// Order matters here: case1 will be preferred to case2.
+#define TRITON_MFMA_v4_2k(m, n, aET, bET, symbol1, k1, kBase1, symbol2, k2,    \
+                          kBase2)                                              \
+  {                                                                            \
+    /*key=*/{4, m, n, aET.getTypeID(), bET.getTypeID()}, /*value=*/{           \
+      {ROCDL::symbol1::getOperationName(), k1, kBase1},                        \
+          {ROCDL::symbol2::getOperationName(), k2, kBase2},                    \
+    }                                                                          \
   }
-  llvm_unreachable("Unsupported input argument type.");
-}
 
-using MfmaInsnGroupMap = llvm::DenseMap<MfmaInsnGroupSelectKey, MfmaInsnAttr,
-                                        MfmaInsnGroupSelectKeyInfo>;
+// Macro for defining MFMA intrinsics existing in multiple gfx versions.
+#define TRITON_MFMA_v1to2(m, n, k, aET, bET, kBase, symbol)                    \
+  TRITON_MFMA_v(1, m, n, k, aET, bET, kBase, symbol),                          \
```

Comment:
The order of args does not match the define in line 78. How does it work?

Reply:
Good catch. It's just C macro--so no type check or whatever. As long as the # of parameters match it is fine; no matter what name they have. I fixed this.